collect_app/src/main/java/org/odk/collect/android/adapters/FileArrayAdapter.java

```diff
@@ -31,57 +31,47 @@
 import org.odk.collect.android.logic.DriveListItem;
 
 import java.text.SimpleDateFormat;
-import java.util.Collections;
-import java.util.Comparator;
 import java.util.Date;
 import java.util.List;
 import java.util.Locale;
 
-public class FileArrayAdapter extends RecyclerView.Adapter<FileArrayAdapter.ViewHolder> {
+import android.support.annotation.NonNull;
+
+public class FileArrayAdapter extends ArrayAdapter<DriveListItem> {
 
     private Context context;
     private List<DriveListItem> items;
 
-    public interface OnItemClickListener {
-        void onItemClick(View v, DriveListItem item);
-    }
-
-    private final OnItemClickListener listener;
 
-    public FileArrayAdapter(Context context, List<DriveListItem> objects, OnItemClickListener listener) {
+    public FileArrayAdapter(Context context, List<DriveListItem> filteredDriveList) {
+        super(context, R.layout.two_item_image, filteredDriveList);
         this.context = context;
-        items = objects;
-        this.listener = listener;
+        items = filteredDriveList;
     }
 
-    static class ViewHolder extends RecyclerView.ViewHolder {
+    private class ViewHolder {
         ImageView imageView;
         TextView text1;
         TextView text2;
         CheckBox checkBox;
-
-        ViewHolder(View v) {
-            super(v);
-            imageView = v.findViewById(R.id.image);
-            text1 = v.findViewById(R.id.text1);
-            text2 = v.findViewById(R.id.text2);
-            checkBox = v.findViewById(R.id.checkbox);
-        }
-
-        void bind(final DriveListItem item, final OnItemClickListener listener) {
-            itemView.setOnClickListener(v -> listener.onItemClick(v, item));
-        }
     }
 
-    @Override
-    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
-        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.two_item_image, parent, false);
-        return new ViewHolder(v);
-    }
-
-    @Override
-    public void onBindViewHolder(ViewHolder holder, int position) {
-        holder.bind(items.get(position), listener);
+    public @NonNull View getView(int position, View convertView, @NonNull ViewGroup parent) {
```

Comment:
`getView` method's responsibility should be only to return the view for the current position. Hence, all the logic related to binding view with data should be held within the view holder class. It is not a rule but a good practice. Something similar to this :

```
@NonNull
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        final ViewHolder holder;
        if (convertView == null) {
            View view = LayoutInflater.from(getContext()).inflate(R.layout.two_item_image, parent, false);
            holder = new ViewHolder(view);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }
        holder.onBind(items.get(position));
        return holder.itemView;
    }
```

Reply:
Then binding of the views can be done in the constructor of the ViewHolder and all the logic related to updating the UI of the view for the current item can be done using `onBind` or some other method rather than exposing the variables

Reply:
```
void onBind(DriveListItem item) {
            setName(item.getName());
            setImage(item.getType());
            setModifiedTime(item.getDate());
            setCheckbox(item);
        }
```

Reply:
The advantage is that, if the listview is replaced by recycler view than the same view holder can be reused as it purely decoupled and just a view holder

Reply:
``` 
private class ViewHolder {
        ImageView imageView;
        TextView text1;
        TextView text2;
        CheckBox checkBox;

        ViewHolder (View view) {
            imageView = view.findViewById(R.id.image);
            text1 = view.findViewById(R.id.text1);
            text2 = view.findViewById(R.id.text2);
            checkBox = view.findViewById(R.id.checkbox);
        }

        void onBind (DriveListItem item) {
            String dateModified = null;
            if (item.getDate() != null) {
                dateModified = new SimpleDateFormat(context.getString(
                        R.string.modified_on_date_at_time), Locale.getDefault())
                        .format(new Date(item.getDate().getValue()));
            }
            // updating the UI elements
       }
}
```
@shobhitagarwal1612 What do you say about this?  Is it required to check whether DriveListItem is null or not?

Reply:
No, it's not